### PR TITLE
synchronize: Use private_key and ssh_user with delegate_to (#16215)

### LINKED ISSUE #16215: delegate_to with synchronize forgets ansible_user
##### ISSUE TYPE
Bug Report

##### COMPONENT NAME
synchronize module

##### ANSIBLE VERSION
```
ansible 2.1.0.0
```

##### CONFIGURATION
host_key_checking = False

##### OS / ENVIRONMENT
Fedora 23

##### SUMMARY
Doing a remote-to-remote push requires using delegate_to. When using synchronize on a host with ansible_user set, and using delegate_to to copy from a second host, Ansible fails to set the remote user for the destination. The remote user is set correctly when delegate_to is removed.

##### STEPS TO REPRODUCE
The task that triggers the bug:

```
      - synchronize:
          src: /home/gemserver/gems/gems
          dest: /var/geminabox/gems/gems
          delete: yes
        delegate_to: "{{ groups['gemserver_master'][0] }}"
```

##### EXPECTED RESULTS
When running with -vvv, the "dest" value should report as "user@host:/var/geminabox/gems/gems",

##### ACTUAL RESULTS
dest is set to just "host:/var/geminabox/gems/gems" (no user@) when delegate_to is present, resulting in being unable to sync.
```
fatal: [111.111.111.111 -> gemserver-master]: FAILED! => {"changed": false, "cmd": "/usr/bin/rsync --delay-updates -F --compress --dry-run --delete-after --archive --rsh 'ssh  -S none -o StrictHostKeyChecking=no' --out-format='<<CHANGED>>%i %n%L' \"/home/gemserver/gems/gems\" \"111.111.111.111:/var/geminabox/gems/gems\"", "failed": true, "invocation": {"module_args": {"_local_rsync_path": "rsync", "_substitute_controller": false, "archive": true, "checksum": false, "compress": true, "copy_links": null, "delete": true, "dest": "111.111.111.111:/var/geminabox/gems/gems", "dest_port": 22, "dirs": false, "existing_only": false, "group": null, "links": null, "mode": "push", "owner": null, "partial": false, "perms": null, "private_key": null, "recursive": null, "rsync_opts": null, "rsync_path": null, "rsync_timeout": 0, "set_remote_user": true, "src": "/home/gemserver/gems/gems", "ssh_args": null, "times": null, "verify_host": false}}, "msg": "Permission denied (publickey,gssapi-keyex,gssapi-with-mic).\r\nrsync: connection unexpectedly closed (0 bytes received so far) [sender]\nrsync error: unexplained error (code 255) at io.c(605) [sender=3.0.9]\n", "rc": 255}
```
(IP address replaced with 111.111.111.111)

### PR DESCRIPTION
**Pull request by Nicop06**
##### SUMMARY

Fixes #16215

When using `delegate_to` with synchronize, the `~` in the private key path is
expanded relative to the machine from which ansible is run and not the host
from which the synchronize is actually run.

Also, the variables `ansible_ssh_user` and `ansible_ssh_private_key_file`
present from the inventory are ignored, leading to inaccessible machine.

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

synchronize

##### ANSIBLE VERSION

```
ansible 2.4.0 (synchronize_module c35fb67ef2) last updated 2017/04/10 19:17:58 (GMT +200)
  config file = /home/nporcel/tools/perf-ansible/ansible.cfg
  configured module search path = [u'/home/nporcel/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/nporcel/tools/ansible/lib/ansible
  executable location = /home/nporcel/.virtualenvs/perf-ansible-local/bin/ansible
  python version = 2.7.12 (default, Jul 18 2016, 15:02:52) [GCC 4.8.4]
```

The problem was also present in ansible 2.2 and 2.3.

##### ADDITIONAL INFORMATION

For instance, let's consider this playbook:

```
- hosts: test
  tasks:
    - name: Send fstab to host
      synchronize:
        src: /etc/fstab
        dest: /tmp
        mode: push
      delegate_to: jumper
```

In the inventory, we have:

```
# test.yml
---
ansible_host: 1.2.3.4
ansible_user: admin
ansible_ssh_private_key_file: ~/.ssh/id_rsa

# jumper.yml
---
ansible_host: 5.6.7.8
ansible_user: jumper
ansible_ssh_private_key_file: ~/.ssh/id_rsa
```

First, without the fix, we get the following rsync command, assuming we are running ansible with user `user`:

```
/usr/bin/rsync (...) -i /home/user/.ssh/id_rsa /etc/fstab 1.2.3.4:/tmp
```

With the fix, the user and ssh key are correctly filled:

```
/usr/bin/rsync (...) -i /home/jumper/.ssh/id_rsa /etc/fstab admin@1.2.3.4:/tmp
```

As you can see, the user is added and the ssh key path is expanded on the jumper instead of the machine where `ansible-playbook` is executed.

### CODE REVIEW

**Comment by mattclay** [comment]
CI failure due to unit test failures (here's one of them, the other appears to be the same):

```
2017-02-20 19:05:58 >           assert value, check
2017-02-20 19:05:58 E           AssertionError: self.final_module_args['dest'] == 'el6host:/tmp/deleteme'
2017-02-20 19:05:58 E           assert False
2017-02-20 19:05:58 
2017-02-20 19:05:58 test/units/plugins/action/test_synchronize.py:178: AssertionError
```

**Comment by mattclay** [comment]
My apologies, the python 3 compatibility issue is unrelated and has been fixed.

**Review by mkrizek — CHANGES_REQUESTED** [review]
@Nicop06 It looks like there has been some work done in the action plugin that is related to your changes here https://github.com/ansible/ansible/pull/25047. Can you check to see if your changes in the action plugin are still needed? If so, can you amend the PR? The changes in the synchronize module itself seems still relevant though.

Thanks!

**Comment by Nicop06** [comment]
@mkrizek You are right, this pull request doesn't work anymore. **Do not merge it yet**, I am working on a better fix.

I even forgot some cases in the original fix such as setup where all 3 machines involved are using a different user.

**Comment by Nicop06** [comment]
I pushed my changes and I updated the description of this PR. Please note that I fix 2 issues with this fix:

* The `ansible_ssh_private_key_file` is expanded on the host referenced by `delegate_to`
* The `ansible_ssh_user` variable of the target host is added to the rsync command.

I tested with and without `delegate_to` and it works in both cases.

**Comment by mattclay** [comment]
CI failure in unit and integration tests: https://app.shippable.com/github/ansible/ansible/runs/34370

**Comment by mattclay** [comment]
CI failure in integration tests:

```
Traceback (most recent call last):
  File "/tmp/ansible_lJH5KV/ansible_module_synchronize.py", line 526, in <module>
    main()
  File "/tmp/ansible_lJH5KV/ansible_module_synchronize.py", line 416, in main
    private_key = os.path.expanduser(private_key)
  File "/usr/lib/python2.7/posixpath.py", line 254, in expanduser
    if not path.startswith('~'):
AttributeError: 'NoneType' object has no attribute 'startswith'
```

**Comment by Nicop06** [comment]
Thank you for the catch. This happens when `private_key` is null so I moved the `expanduser` call after the validity check.

**Comment by Nicop06** [comment]
The errors are from the integration tests of the `uri` module which should not be impacted by my changes. It seems to be an environment issue anyway.

File: lib/ansible/plugins/action/synchronize.py [hunk 1]
```diff
@@ -326,7 +328,6 @@ def run(self, tmp=None, task_vars=None):
             private_key = self._play_context.private_key_file
 
             if private_key is not None:
-                private_key = os.path.expanduser(private_key)
```
Comment (mkrizek) [inline lib/ansible/plugins/action/synchronize.py:329]: Is this change really needed? Removing `os.path.expanduser` here and calling it in the module?

**Review by mkrizek — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/action/synchronize.py (hunk 1 above)
Comment (Nicop06) [reply lib/ansible/plugins/action/synchronize.py:329]: This will allow to expand the home directory on the delegated host and not the machine running ansible, as the module is run on the delegate host while the action is run in local.

The effect is your private key is set to `~/.ssh/myhost.key`, the `~` will be expanded to the home directory of the user used to connect to the delegate host instead of the home directory of the machine where you are running the playbook.

You can see the effect around the end of the description of this PR.

Comment (mkrizek) [reply lib/ansible/plugins/action/synchronize.py:329]: Oh, sorry, got it. Should we have `expanduser` in both places to be "defensive", in case `private_user` is used somewhere else in the module in the future?

Comment (Nicop06) [reply lib/ansible/plugins/action/synchronize.py:329]: The fix only works if `expanduser` is not called from the action. In fact, from what I understand in that case, the module is sent to the delegate host and the action builds the parameters that will be used for the call in the machine running ansible.

So calling `expanduser` both from the action and the module will expand it on the local machine and then on the delegate. But because it was already expanded on the local machine, expanding it on the delegate won't have any effect and the path will be relative to the local machine (the one running ansible).

In my test, I checked that it works well with and without delegate, so I don't expect any regression with this patch. Also, you make a valid point and we could remove the `private_key` variable and directly do:

```
_tmp_args['private_key'] = self._play_context.private_key_file
```

Without even checking if the key is none as we already check that on the module. If you prefer this fix, I can update my PR.

**Review by mkrizek — APPROVED** [review]

**Comment by dagwieers** [comment]
So I am not convinced that this is the correct fix. (i.e. remote_user affects the target, not the delegated host, it's always been like this)

And if this would be the correct fix, it should be fixed at another level than the synchronize action plugin. you now have made the **synchronize** module behave differently than other modules.

cc: @bcoca @jctanner

**Comment by Nicop06** [comment]
Actually, the **synchronize** module is slightly different from other modules. The `rsync` command is run directly on the machine running `ansible` instead of the target machine. When using a delegate, it runs `rsync` on the delegate host with the target host as destination and it requires the ssh parameters of the target host to establish the connection. As far as I know, this is the only module in the core ansible modules using the ssh parameters of the target host when running with `delegate_to`.

But maybe the problem could be solved, for example by setting an explicit `dest` parameter including the correct user and host in the playbook itself. Although, it might make the option `set_remote_user` obsolete if not confusing.

That being said, I am happy if someone has a fresher look to this problem and can find a better way to solve it.
